File: IPAddress/src/inet.ipaddr/inet/ipaddr/format/util/Partition.java
/*
 * Copyright 2020 Sean C Foley
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *     or at
 *     https://github.com/seancfoley/IPAddress/blob/master/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package inet.ipaddr.format.util;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;

import inet.ipaddr.Address;
import inet.ipaddr.AddressSegmentSeries;
import inet.ipaddr.IPAddress;
import inet.ipaddr.IPAddressSegmentSeries;

/**
 * Represents a partition of an element, such as a subnet,
 * into one or elements of the same type that represent the same set of values.
 * <p>
 * For instance, it can represent the partition of a subnet into prefix blocks, or a partition into individual addresses.
 * <p>
 * You can use the methods {@link #partitionWithSingleBlockSize(AddressSegmentSeries)} or {@link #partitionWithSpanningBlocks(IPAddress)}
 * to create a partition, or you can create your own.
 * <p>
 * You can use this class with the trie classes to easily partition an address for applying an operation to the trie.
 * Given an address or subnet s, you can partition into addresses or blocks that can used with a trie class using
 * <code>Partition.partitionWithSpanningBlocks(s)</code> or <code>Partition.partitionWithSingleBlockSize(s)</code>.
 * Once you have your partition, you choose a method in the Partition class whose argument type matches the signature of the trie method you wish to use.
 * <p>
 * For instance, you can use methods like {@link AddressTrieOps#contains(inet.ipaddr.Address)} or {@link AddressTrieOps#remove(inet.ipaddr.Address)} with {@link Partition#predicateForAny(Predicate)} or {@link Partition#predicateForEach(Predicate)} 
 * since the contains or remove method matches the Predicate interface.
 * Methods that return non-boolean values would match the {@link Partition#applyForEach(Function)} or {@link Partition#forEach(Consumer)} methods, as in the following code example for a given subnet s of type E:
 * <code>Map&lt;E, TrieNode&lt;E&gt;&gt; all = Partition.partitionWithSingleBlockSize(s).applyForEach(trie::getNode)</code>
 *
 * @author scfoley
 *
 * @param <E> the type being partitioned
 */
public class Partition<E> {
	/**
	 * The partitioned address.
	 */
	public final E original;
	
	/**
	 * A field containing a partition into a single value.
	 * <p>
	 * When {@link #blocks} is null, 
	 * the partition result is stored in this field.
	 * This will match {@link #original} in terms of equality,
	 * but might not have the same prefix length.
	 * 
	 */
	public final E single;
	
	/**
	 * An iterator supplying the partitioned addresses.
	 * <p>
	 * When {@link #single} is null, the partition result is stored in this field.
	 * <p>
	 * If the partition result is multiple addresses or blocks, 
	 * they are supplied by this iterator.
	 * <p>
	 * If the partition result is just a single address, 
	 * the result may be supplied by this iterator, or optionally this iterator may be null,
	 * in which case the result is given by {@link #single}.
	 */
	public final Iterator<? extends E> blocks;
	
	/**
	 * The address or block count.
	 * <p>
	 * If the result is greater than 1, the blocks are supplied by {@link #blocks}.
	 * If the result is 1, then the single block or address is supplied by {@link #blocks} if it is non-null,
	 * otherwise the single block or address is in {@link #single}.
	 */
	public final BigInteger count;
	
	public Partition(E original) {
		this.count = BigInteger.ONE;
		this.single = original;
		this.blocks = null;
		this.original = original;
	}
	
	public Partition(E original, E single) {
		this.count = BigInteger.ONE;
		this.single = single;
		this.blocks = null;
		this.original = original;
	}
	
	public Partition(E original, Iterator<? extends E> blocks, BigInteger count) {
		this.count = count;
		this.single = null;
		this.blocks = blocks;
		this.original = original;
	}
	
	
	public Partition(E original, Iterator<? extends E> blocks, int count) {
		this.count = BigInteger.valueOf(count);
		this.single = null;
		this.blocks = blocks;
		this.original = original;
	}

	/**
	 * Supplies to the given function each element of this partition,
	 * inserting non-null return values into the returned map.
	 * 
	 * @param <R>
	 * @param func
	 * @return
	 */
	public <R> Map<E, R> applyForEach(Function<? super E, ? extends R> func) {
		TreeMap<E, R> results = new TreeMap<>();
		forEach(address -> {
			R result = func.apply(address);
			if(result != null) {
				results.put(address, result);
			}
		});
		return results;
	}

	/**
	 * Supplies to the consumer each element of this partition.
	 * 
	 * @param action
	 */
	public void forEach(Consumer<? super E> action) {
		if(blocks == null) {
			action.accept(single);
		} else {
			Iterator<? extends E> iterator = blocks;
			while(iterator.hasNext()) {
				action.accept(iterator.next());
			}
		}
	}

	/**
	 * Applies the operation to each element of the partition,
	 * returning true if they all return true, false otherwise
	 * @param predicate
	 * @return
	 */
	public boolean predicateForEach(Predicate<? super E> predicate) {
		return predicateForEach(predicate, false);
	}

	/**
	 * Applies the operation to each element of the partition,
	 * returning true if they all return true, false otherwise
	 * @param predicate
	 * @param returnEarly returns as soon as one application of the predicate returns false (determining the overall result)
	 * @return
	 */
	public boolean predicateForEach(Predicate<? super E> predicate, boolean returnEarly) {
		if(blocks == null) {
			return predicate.test(single);
		}
		boolean result = true;
		Iterator<? extends E> iterator = blocks;
		while(iterator.hasNext()) {
			if(!predicate.test(iterator.next())) {
				result = false;
				if(returnEarly) {
					break;
				}
			}
		}
		return result;
	}

	/**
	 * Applies the operation to each element of the partition, 
	 * returning true if the given predicate returns true for any of the elements.
	 * 
	 * @param predicate
	 * @param returnEarly returns as soon as one call to the predicate returns true
	 * @return
	 */
	public boolean predicateForAny(Predicate<? super E> predicate, boolean returnEarly) {
		return !predicateForEach((addr) -> !predicate.test(addr), returnEarly);
	}

	/**
	 * Applies the operation to each element of the partition, 
	 * returning true if the given predicate returns true for any of the elements.
	 * 
	 * @param predicate
	 * @return
	 */
	public boolean predicateForAny(Predicate<? super E> predicate) {
		return predicateForAny(predicate, false);
	}

	/**
	 * Partitions the address series into prefix blocks and single addresses.
	 * <p>
	 * If null is returned, the argument is already an individual address or prefix block.
	 * <p>
	 * This method iterates through a list of prefix blocks of different sizes that span the entire subnet.
	 * 
	 * @param newAddr
	 * @return
	 */
	@SuppressWarnings("unchecked")
	public static <E extends IPAddress> Partition<E> partitionWithSpanningBlocks(E newAddr) {
		if(!newAddr.isMultiple()) {
			if(!newAddr.isPrefixed()) {
				return new Partition<E>(newAddr);
			}
			return new Partition<E>(newAddr, (E) newAddr.withoutPrefixLength());
		} else if(newAddr.isSinglePrefixBlock()) {
			return new Partition<E>(newAddr);
		}
		E blocks[] = (E[]) newAddr.spanWithPrefixBlocks();
		Iterator<? extends E> blocksIterator = Arrays.asList(blocks).iterator();
		return new Partition<E>(newAddr, blocksIterator, blocks.length);
	}

	/**
	 * Partitions the address series into prefix blocks and single addresses.
	 * <p>
	 * If null is returned, the argument is already an individual address or prefix block.
	 * <p>
	 * This method chooses the maximum block size for a list of prefix blocks contained by the address or subnet,
	 *  and then iterates to produce blocks of that size.
	 * 
	 * @param newAddr
	 * @return
	 */
	@SuppressWarnings("unchecked")
	public static <E extends AddressSegmentSeries> Partition<E> partitionWithSingleBlockSize(E newAddr) {
		if(!newAddr.isMultiple()) {
			if(!newAddr.isPrefixed()) {
				return new Partition<E>(newAddr);
			}
			return new Partition<E>(newAddr, (E) newAddr.withoutPrefixLength());
		} else if(newAddr.isSinglePrefixBlock()) {
			return new Partition<E>(newAddr);
		}
		// prefix blocks are handled as prefix blocks, 
		// such as 1.2.*.*, which is handled as prefix block iterator for 1.2.0.0/16,
		// but 1.2.3-4.5 is handled as iterator with no prefix lengths involved
		if(newAddr instanceof IPAddressSegmentSeries) {
			IPAddressSegmentSeries series = ((IPAddressSegmentSeries) newAddr).assignMinPrefixForBlock();
			if(series.getPrefixLength() != newAddr.getBitCount()) {
				Iterator<? extends E> iterator = (Iterator<? extends E>) series.prefixBlockIterator();
				return new Partition<E>(newAddr, iterator, series.getPrefixCount(series.getPrefixLength()));
			}
		} else {
			int prefLen = newAddr.getMinPrefixLengthForBlock();
			if(prefLen != newAddr.getBitCount()) {
				AddressSegmentSeries series = newAddr.setPrefixLength(prefLen, false);
				Iterator<? extends E> iterator = (Iterator<? extends E>) series.prefixBlockIterator();
				return new Partition<E>(newAddr, iterator, series.getPrefixCount(series.getPrefixLength()));
			}
		}
		Iterator<? extends E> iterator = (Iterator<? extends E>) newAddr.withoutPrefixLength().iterator();
		return new Partition<E>(newAddr, iterator, newAddr.getCount());
	}

	/**
	 * If the given address is a single prefix block, it is returned.
	 * If it can be converted to a single prefix block or address (by adjusting the prefix length), the converted block is returned.
	 * Otherwise, null is returned.
	 * 
	 * @param <E>
	 * @param addr
	 * @return
	 */
	public static <E extends Address> E checkBlockOrAddress(E addr) {
		 return AbstractTree.checkBlockOrAddress(addr, false);
	}
}

File: ipaddress-go/ipaddr/partition.go
//
// Copyright 2020-2021 Sean C Foley
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package ipaddr

import (
	"github.com/seancfoley/ipaddress/ipaddress-go/ipaddr/addrerr"
	"math/big"
)

// A partition is a collection of addresses partitioned from an original address.
// Much like an iterator, the elements of the partition can be iterated just once, after which it becomes empty.
type Partition struct {
	original *IPAddress

	single *IPAddress

	iterator IPAddressIterator

	count *big.Int
}

type IPv6Partition struct {
	p *Partition
}

func (p IPv6Partition) ForEach(action func(*IPv6Address)) {
	p.p.ForEach(func(address *IPAddress) {
		action(address.ToIPv6())
	})
}

func (p IPv6Partition) Iterator() IPv6AddressIterator {
	return ipv6IPAddressIterator{p.p.Iterator()}
}

// PredicateForEach applies the operation to each element of the partition,
// returning true if they all return true, false otherwise
func (p IPv6Partition) PredicateForEach(predicate func(*IPv6Address) bool) bool {
	return p.p.PredicateForEach(func(address *IPAddress) bool {
		return predicate(address.ToIPv6())
	})
}

// PredicateForEachEarly applies the operation to each element of the partition,
// returning false if the given predicate returns false for any of the elements.
//
// The method returns when one application of the predicate returns false (determining the overall result)
func (p IPv6Partition) PredicateForEachEarly(predicate func(*IPv6Address) bool) bool {
	return p.p.PredicateForEachEarly(func(address *IPAddress) bool {
		return predicate(address.ToIPv6())
	})
}

// PredicateForAnyEarly applies the operation to each element of the partition,
// returning true if the given predicate returns true for any of the elements.
//
// The method returns when one application of the predicate returns true (determining the overall result)
func (p IPv6Partition) PredicateForAnyEarly(predicate func(*IPv6Address) bool) bool {
	return p.p.PredicateForAnyEarly(func(address *IPAddress) bool {
		return predicate(address.ToIPv6())
	})
}

// PredicateForAny applies the operation to each element of the partition,
// returning true if the given predicate returns true for any of the elements.
func (p IPv6Partition) PredicateForAny(predicate func(*IPv6Address) bool) bool {
	return p.p.PredicateForAny(func(address *IPAddress) bool {
		return predicate(address.ToIPv6())
	})
}

type IPv4Partition struct {
	p *Partition
}

func (p IPv4Partition) ForEach(action func(*IPv4Address)) {
	p.p.ForEach(func(address *IPAddress) {
		action(address.ToIPv4())
	})
}

func (p IPv4Partition) Iterator() IPv4AddressIterator {
	return ipv4IPAddressIterator{p.p.Iterator()}
}

// PredicateForEach applies the operation to each element of the partition,
// returning true if they all return true, false otherwise
func (p IPv4Partition) PredicateForEach(predicate func(*IPv4Address) bool) bool {
	return p.p.PredicateForEach(func(address *IPAddress) bool {
		return predicate(address.ToIPv4())
	})
}

// PredicateForEachEarly applies the operation to each element of the partition,
// returning false if the given predicate returns false for any of the elements.
//
// The method returns when one application of the predicate returns false (determining the overall result)
func (p IPv4Partition) PredicateForEachEarly(predicate func(*IPv4Address) bool) bool {
	return p.p.PredicateForEachEarly(func(address *IPAddress) bool {
		return predicate(address.ToIPv4())
	})
}

// PredicateForAnyEarly applies the operation to each element of the partition,
// returning true if the given predicate returns true for any of the elements.
//
// The method returns when one application of the predicate returns true (determining the overall result)
func (p IPv4Partition) PredicateForAnyEarly(predicate func(*IPv4Address) bool) bool {
	return p.p.PredicateForAnyEarly(func(address *IPAddress) bool {
		return predicate(address.ToIPv4())
	})
}

// PredicateForAny applies the operation to each element of the partition,
// returning true if the given predicate returns true for any of the elements.
func (p IPv4Partition) PredicateForAny(predicate func(*IPv4Address) bool) bool {
	return p.p.PredicateForAny(func(address *IPAddress) bool {
		return predicate(address.ToIPv4())
	})
}

// TODO LATER this needs generics (of course, the whole Partition type could be generic as well)
// Supplies to the given function each element of this partition,
// inserting non-null return values into the returned map.
//	public <R> Map<E, R> ApplyForEach(Function<? super E, ? extends R> func) map[]{
//		TreeMap<E, R> results = new TreeMap<>();
//		forEach(address -> {
//			R result = func.apply(address);
//			if(result != null) {
//				results.put(address, result);
//			}
//		});
//		return results;
//	}

// Supplies to the consumer each element of this partition.
func (p *Partition) ForEach(action func(*IPAddress)) {
	if p.iterator == nil {
		item := p.single
		if item != nil {
			p.single = nil
			action(item)
		}
	} else {
		iterator := p.iterator
		for iterator.HasNext() {
			action(iterator.Next())
		}
		p.iterator = nil
	}
}

func (p *Partition) Iterator() IPAddressIterator {
	if p.iterator == nil {
		item := p.single
		if item != nil {
			res := &ipAddrIterator{&singleAddrIterator{item.ToAddressBase()}}
			item = nil
			return res
		}
		return nil
	}
	res := p.iterator
	p.iterator = nil
	return res
}

// Applies the operation to each element of the partition,
// returning true if they all return true, false otherwise
func (p *Partition) PredicateForEach(predicate func(*IPAddress) bool) bool {
	return p.predicateForEach(predicate, false)
}

// Applies the operation to each element of the partition,
// returning false if the given predicate returns false for any of the elements.
//
// The method returns when one application of the predicate returns false (determining the overall result)
func (p *Partition) PredicateForEachEarly(predicate func(*IPAddress) bool) bool {
	return p.predicateForEach(predicate, false)
}

func (p *Partition) predicateForEach(predicate func(*IPAddress) bool, returnEarly bool) bool {
	if p.iterator == nil {
		return predicate(p.single)
	}
	result := true
	iterator := p.iterator
	for iterator.HasNext() {
		if !predicate(iterator.Next()) {
			result = false
			if returnEarly {
				break
			}
		}
	}
	return result
}

// Applies the operation to each element of the partition,
// returning true if the given predicate returns true for any of the elements.
//
// The method returns when one application of the predicate returns true (determining the overall result)
func (p *Partition) PredicateForAnyEarly(predicate func(*IPAddress) bool) bool {
	return p.predicateForAny(predicate, true)
}

// Applies the operation to each element of the partition,
// returning true if the given predicate returns true for any of the elements.
func (p *Partition) PredicateForAny(predicate func(*IPAddress) bool) bool {
	return p.predicateForAny(predicate, false)
}

func (p *Partition) predicateForAny(predicate func(address *IPAddress) bool, returnEarly bool) bool {
	return !p.predicateForEach(func(addr *IPAddress) bool {
		return !predicate(addr)
	}, returnEarly)
}

// PartitionWithSpanningBlocks partitions the address series into prefix blocks and single addresses.
//
// This method iterates through a list of prefix blocks of different sizes that span the entire subnet.
func PartitionIpv6WithSpanningBlocks(newAddr *IPv6Address) IPv6Partition {
	return IPv6Partition{partitionWithSpanningBlocks(newAddr.ToIP())}
}

// PartitionWithSpanningBlocks partitions the address series into prefix blocks and single addresses.
//
// This method iterates through a list of prefix blocks of different sizes that span the entire subnet.
func PartitionIpv4WithSpanningBlocks(newAddr *IPv4Address) IPv4Partition {
	return IPv4Partition{partitionWithSpanningBlocks(newAddr.ToIP())}
}

// PartitionWithSpanningBlocks partitions the address series into prefix blocks and single addresses.
//
// This method iterates through a list of prefix blocks of different sizes that span the entire subnet.
func partitionWithSpanningBlocks(newAddr *IPAddress) *Partition {
	if !newAddr.isMultiple() {
		if !newAddr.IsPrefixed() {
			return &Partition{
				original: newAddr,
				single:   newAddr,
				count:    bigOneConst(),
			}
		}
		return &Partition{
			original: newAddr,
			single:   newAddr.WithoutPrefixLen(),
			count:    bigOneConst(),
		}
	} else if newAddr.IsSinglePrefixBlock() {
		return &Partition{
			original: newAddr,
			single:   newAddr,
			count:    bigOneConst(),
		}
	}
	blocks := newAddr.SpanWithPrefixBlocks()
	return &Partition{
		original: newAddr,
		iterator: ipAddrSliceIterator{blocks},
		count:    big.NewInt(int64(len(blocks))),
	}
}

// PartitionIPv6WithSingleBlockSize partitions the address series into prefix blocks and single addresses.
//
// This method chooses the maximum block size for a list of prefix blocks contained by the address or subnet,
// and then iterates to produce blocks of that size.
func PartitionIPv6WithSingleBlockSize(newAddr *IPv6Address) IPv6Partition {
	return IPv6Partition{partitionWithSingleBlockSize(newAddr.ToIP())}
}

// PartitionIPv4WithSingleBlockSize partitions the address series into prefix blocks and single addresses.
//
// This method chooses the maximum block size for a list of prefix blocks contained by the address or subnet,
// and then iterates to produce blocks of that size.
func PartitionIPv4WithSingleBlockSize(newAddr *IPv4Address) IPv4Partition {
	return IPv4Partition{partitionWithSingleBlockSize(newAddr.ToIP())}
}

// PartitionWithSingleBlockSize partitions the address series into prefix blocks and single addresses.
//
// This method chooses the maximum block size for a list of prefix blocks contained by the address or subnet,
// and then iterates to produce blocks of that size.
func partitionWithSingleBlockSize(newAddr *IPAddress) *Partition {
	if !newAddr.isMultiple() {
		if !newAddr.IsPrefixed() {
			return &Partition{
				original: newAddr,
				single:   newAddr,
				count:    bigOneConst(),
			}
		}
		return &Partition{
			original: newAddr,
			single:   newAddr.WithoutPrefixLen(),
			count:    bigOneConst(),
		}
	} else if newAddr.IsSinglePrefixBlock() {
		return &Partition{
			original: newAddr,
			single:   newAddr,
			count:    bigOneConst(),
		}
	}
	// prefix blocks are handled as prefix blocks,
	// such as 1.2.*.*, which is handled as prefix block iterator for 1.2.0.0/16,
	// but 1.2.3-4.5 is handled as iterator with no prefix lengths involved
	series := newAddr.AssignMinPrefixForBlock()
	if series.getPrefixLen().bitCount() != newAddr.GetBitCount() {
		return &Partition{
			original: newAddr,
			iterator: series.PrefixBlockIterator(),
			count:    series.GetPrefixCountLen(series.getPrefixLen().bitCount()),
		}
	}
	return &Partition{
		original: newAddr,
		iterator: newAddr.WithoutPrefixLen().Iterator(),
		count:    newAddr.GetCount(),
	}
}

// CheckBlockOrAddress converts to a single prefix block or address.
// If the given address is a single prefix block, it is returned.
// If it can be converted to a single prefix block or address (by adjusting the prefix length), the converted block is returned.
// Otherwise, nil is returned.
func CheckBlockOrAddress(addr *IPAddress) *IPAddress {
	res, _ := checkBlockOrAddress(addr)
	return res
}

// Ensures the address is either an individual address or a prefix block subnet.
func checkBlockOrAddress(addr *IPAddress) (*IPAddress, addrerr.IncompatibleAddressError) {
	if !addr.isMultiple() {
		if !addr.IsPrefixed() {
			return addr, nil
		}
		return addr.WithoutPrefixLen(), nil
	} else if addr.IsSinglePrefixBlock() {
		return addr, nil
	} else {
		series := addr.AssignPrefixForSingleBlock()
		if series != nil {
			return series, nil
		}
	}
	return nil, &incompatibleAddressError{addressError{key: "ipaddress.error.address.not.block"}}
}
